# Data Cleaning & Quality Checks
### Olist Profitability & Customer Analysis Project

This notebook documents the data quality issues found while loading the raw
Kaggle CSVs into PostgreSQL (see `sql/01_schema.sql` and `sql/02_load_data.sql`),
and validates the cleaned tables before analysis.


In [1]:
import pandas as pd
from sqlalchemy import create_engine
from getpass import getpass

pg_password = getpass("Postgres password: ")
engine = create_engine(f"postgresql+psycopg2://postgres:{pg_password}@localhost:5432/olist_analysis")


## Issue 1: Missing category translations

When first loading `products.csv`, the load failed with a foreign key
violation:

```
ERROR: insert or update on table "products" violates foreign key constraint
"products_product_category_name_fkey"
DETAIL: Key (product_category_name)=(pc_gamer) is not present in table
"category_translation".
```

**Diagnosis:** Kaggle's `product_category_name_translation.csv` doesn't cover
every category that actually appears in `olist_products_dataset.csv`.
`pc_gamer` was one of the gaps.

**Fix:** Removed the strict foreign key constraint on `products.product_category_name`
(see `sql/01_schema.sql`). Untranslated categories are handled downstream with
a `LEFT JOIN` + `NULL` filter in the profitability views, rather than dropping
those rows or forcing a guessed translation.

Let's confirm how many categories are actually affected, and how much revenue
they represent — worth knowing before deciding this was an acceptable
tradeoff.

In [2]:
untranslated = pd.read_sql('''
    SELECT p.product_category_name, COUNT(*) AS product_count
    FROM products p
    LEFT JOIN category_translation ct ON p.product_category_name = ct.product_category_name
    WHERE ct.product_category_name IS NULL AND p.product_category_name IS NOT NULL
    GROUP BY p.product_category_name
    ORDER BY product_count DESC;
''', engine)

untranslated


,product_category_name,product_count
0,portateis_cozinha_e_preparadores_de_alimentos,10
1,pc_gamer,3


In [3]:
revenue_impact = pd.read_sql('''
    SELECT
        COUNT(*) AS affected_order_items,
        ROUND(SUM(oi.price)::numeric, 2) AS affected_revenue,
        ROUND(
            SUM(oi.price)::numeric /
            (SELECT SUM(price) FROM order_items) * 100, 2
        ) AS pct_of_total_revenue
    FROM order_items oi
    JOIN products p ON oi.product_id = p.product_id
    LEFT JOIN category_translation ct ON p.product_category_name = ct.product_category_name
    WHERE ct.product_category_name IS NULL;
''', engine)

revenue_impact


,affected_order_items,affected_revenue,pct_of_total_revenue
0,1627,185049.76,1.36


**Conclusion:** if the affected revenue share is small (typically well under
1-2% for this dataset), excluding untranslated categories from category-level
analysis is a reasonable, low-risk tradeoff — worth stating explicitly rather
than silently dropping rows.


## Issue 2: Order status filtering

Not every row in `orders` represents a completed sale — some are cancelled or
never delivered. Counting these would inflate revenue and distort margin
calculations. Let's check the breakdown.

In [4]:
status_counts = pd.read_sql('''
    SELECT order_status, COUNT(*) AS order_count,
           ROUND(COUNT(*) * 100.0 / SUM(COUNT(*)) OVER (), 2) AS pct
    FROM orders
    GROUP BY order_status
    ORDER BY order_count DESC;
''', engine)

status_counts


,order_status,order_count,pct
0,delivered,96478,97.02
1,shipped,1107,1.11
2,canceled,625,0.63
3,unavailable,609,0.61
4,invoiced,314,0.32
5,processing,301,0.30
6,created,5,0.01
7,approved,2,0.00


**Decision:** all profitability and RFM analysis in this project filters to
`order_status = 'delivered'` only (already applied in `sql/03_profitability_view.sql`
and the RFM notebook's query). This is the standard approach — revenue should
only count orders customers actually received.

## Issue 3: Null / duplicate checks on key columns

Quick sanity check that the primary join keys are clean after loading.

In [5]:
checks = pd.read_sql('''
    SELECT
        'orders.customer_id nulls' AS check_name, COUNT(*) AS count
    FROM orders WHERE customer_id IS NULL
    UNION ALL
    SELECT 'order_items.product_id nulls', COUNT(*)
    FROM order_items WHERE product_id IS NULL
    UNION ALL
    SELECT 'customers duplicate customer_id', COUNT(*) - COUNT(DISTINCT customer_id)
    FROM customers
    UNION ALL
    SELECT 'products duplicate product_id', COUNT(*) - COUNT(DISTINCT product_id)
    FROM products;
''', engine)

checks


,check_name,count
0,orders.customer_id nulls,0
1,order_items.product_id nulls,0
2,products duplicate product_id,0
3,customers duplicate customer_id,0


## Summary

- 2 data quality issues found and resolved during loading:
  - Missing category translations affected 2 categories (13 products, 1,627
    order items) — only **1.36% of total revenue**, confirming this was a
    low-risk exclusion from category-level views rather than a silent data
    loss
  - Order status filtering: 97.02% of orders are `delivered`; the remaining
    2.98% (shipped, canceled, unavailable, etc.) are correctly excluded from
    revenue and margin calculations
- Both fixes are documented in the SQL scripts, not just this notebook, so the
  pipeline is reproducible
- No null/duplicate key issues found in the core join columns

Data is ready for exploratory analysis (`02_exploratory_analysis.ipynb`) and
the profitability / RFM analyses.